man/dbmerge.1
.TH DBMERGE 1 " 4 April 1996"
.FB
.SH NAME
dbmerge \- merge source text with a \fIFirstBase\fP database (mail merge)
.SH SYNOPSIS
.B dbmerge
[\fB-b\fP] [\fB-y\fP] [\fB-B\fP]
[\fB-n\fP \fIcount\fP]
[\fB-w\fP \fIwidth\fP]
[\fB-d\fP \fIdatabase\fP]
[\fB-i\fP \fIindex\fP]
[\fIsource\fP]
[\fIdestination\fP | \fB-\fP]
.SH DESCRIPTION
.I Dbmerge
merges a source document with user selected fields from a \fIFirstBase\fP
database.
The end result is a separate copy of the input document for each record
in the current \fIFirstBase\fP index with actual database data substituted for
in indicated places.
The set of processed records can be limited using the \fB-i\fP index specifier.
The \fB-n\fP flag can be used to process exactly \fIcount\fP records.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell
scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
The \fB-B\fP flag will enable the fastest throughput of records from the
database. If this flag can be used, the \fIdbmerge\fP generation
is speeded up a great deal by internal record buffering.
Note that no index can be used with record buffering, and
that record locking is disabled.
.PP
The \fB-w\fP flag is used to set the width of the document. Lines longer
than this width will be truncated with a warning.
The default width is the \fIFirstBase\fP environment variable \fIMERGECOLS\fP
or the minimum width of a \fIFirstBase\fP controlling terminal, 80.
.PP
If command line arguments for the source or destination are
not provided, dbmerge will prompt for
the
.I source 
and/or
.I destination
file names.
The destination file may be specified as \fB-\fP to produce output on the
standard output device.
.PP
To begin, a document must be prepared using some kind of
.I text editor,
or by using \fIdbdmrg\fP(1).
This document is the
.I source
to 
\fIdbmerge\fP(1).
One good example of 
\fIdbmerge\fP(1)
input is an
\fInroff\fP(1)
source document.
.PP
The input document may be interspersed with 
.I tokens
that represent data fields in a \fIFirstBase\fP database. These tokens are
formed by pre-pending a dollar sign, \fB$\fP,
to a valid \fIFirstBase\fP field_name.
For example, if 
.I LastName
is a valid database field name as defined in the current database
dictionary, then
.B $LastName
would be the correct token to insert into the document.
Again, this prepared document is the input or 
.I source
to
\fIdbmerge\fP(1).
.PP
Tokens can also be forced to appear in an expanded or truncated manner.
Each token can be followed by a size delimiter. For example,
\fB$LastName[25]\fP would print the LastName contents in a slot exactly
25 characters wide, no matter how big the actual data field is.
A size of 0 or no size means the field token data will print unconstrained.
.PP
\fIDbmerge\fP(1)
first parses the 
.I source
looking for tokens to be expanded.
During this stage, all found tokens are displayed on the screen.
Unresolved tokens are marked 
.I ...undefined. 
These undefined tokens are
printed out in place exactly as they appear.
Thus, '$100.00' would be marked as undefined, and would print '$100.00'.
Note: to get a single dollar sign without a warning, use two dollar signs ($$).
.PP
After getting the user's permission to continue, 
the processing stage takes the 
.I source 
document and produces one
copy for each record in the \fIFirstBase\fP index to the 
.I destination 
file.
Defined tokens are substituted for their respective field data from the
database during this stage.
.PP
.I Note:
white spaces (tabs, blanks, newlines) are kept intact during the actual
document-database merging.
Furthermore, it is up to the user to prepare the document properly.
For example, if the document is a letter, then probably a new page
is needed for each letter produced. For nroff source, this
means that a \fB.bp\fP command must be either the first or last nroff
command in the document.
.PP
The source document can also contain specific sections of text that are
processed exactly once: at the beginning or at the end of the
entire merge process. For example, this feature could be used to place
a header at the top of some merge output.
.PP
To use these features, the symbols \fB_BEGIN_\fP and \fB_END_\fP are used
to delimit sections of text to be processed at the beginning or at the end
of the merge process.
Both sections are optional, but if used,
should appear as in the following example:
.sp 1
.nf
.nj
_BEGIN_
...
[ Header information \(em printed exactly once ]
...
_BEGIN_
...
[ Normal Text ... with embedded $ Field Tokens ]
...
_END_
[ Footer information \(em printed exactly once ]
_END_
.ju
.fi
.sp 1
.SH FILES
.PD 0
.TP 10
merge.in
default name for merge source file.
.TP 10
merge.out
default name for the merge output file.
.TP 10
dbase
default name for the \fIFirstBase\fP database.
.TP 10
index
default name for the \fIFirstBase\fP index.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.idx
\fIFirstBase\fP index.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary.
.PD
.SH SEE ALSO
dbdmrg(1), dbedit(1), nroff(1), generators(5), firstbase(5), setup(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbmerge\fP(1)
pauses to ask permission to destroy an existing output
(\fIdestination\fP) file.
A valid \fIFirstBase\fP database dictionary is required for the current
database.
.sp 1
This tool uses a \fIFirstBase\fP index to generate the output from the
database in a particular order.
If this index is detected to be out of date or non-existent, it is
regenerated \fIautomatically\fP. An index is out of date if 
either the database or the index dictionary have changed more
recently than the index was generated.
.sp 1
This tool is line oriented and expects lines of 250 characters or less.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbmerge/dbmerge.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbmerge.c,v 9.1 2001/02/16 19:44:23 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbmerge.c	8.5 04 Apr 1996 FB";
#endif

/*
 * dbmerge.c - provides an interface between nroff (etc) and
 *   the cdb system. field_name tokens on the fb_input stream are replaced
 *   before being written to the output stream by the fields
 *   of a single record. the fb_input stream is rewound, and
 *   the process started again with a new record (the next in the index).
 *
 *   this is the cdb version of mail-merge systems.
 */

#include <dbdmrg_v.h>

#define SCREENLIMIT 15
#define BEGINSYMBOL	"_BEGIN_"
#define ENDSYMBOL	"_END_"

static char inpname[FB_MAXNAME] = {""};	/* input file name */

int counter;				/* record counter */
int outfd;				/* output file descriptor */
int ncount = 0;				/* -n ncount records only flag */
int Bflag = 0;				/* -B blocking factor flag */

fb_aline *a_begin = NULL;
fb_aline *a_end = NULL;

extern short int cdb_mergecols;
extern short int cdb_interrupt;
extern short int cdb_screenprint;

static initm();
static mparse();
static merge();
static mkpart();
static checkbegin();

/*
 *  merge - main driver
 */
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char *buf, filen[FB_MAXNAME];

	 if (fb_testargs(argc, argv, "-") > 0)
	    cdb_batchmode = 1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 if (fb_testargs(argc, argv, "-i") > 0)
	    fb_opendb(cdb_db, READ, FB_WITHINDEX, FB_MUST_INDEX);
         else
	    fb_opendb(cdb_db, READ, FB_NOINDEX, FB_MUST_INDEX);
         fb_scrhdr(cdb_db, "Parsing"); fb_infoline();
         initm(argc, argv, filen);
         buf = (char *) fb_malloc(MAX(hp->recsiz,hp->irecsiz)+10);
	 if (!cdb_yesmode && !cdb_batchmode)
            mparse(filen);
	 if (!cdb_yesmode && !cdb_batchmode){
            if (fb_mustbe('y',
	             "If accurate, enter 'y' to continue:",
                     cdb_t_lines, 1) == FB_ERROR)
               fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    fb_move(3, 1), fb_clrtobot(), fb_infoline();
	    }
	 if (cdb_interrupt)
	    fb_allow_int();
         if (merge(buf, filen) == FB_ERROR)
            fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 fb_closedb(cdb_db);
	 if (!cdb_batchmode && cdb_screenprint)
	    fb_screenprint(filen);
         fb_ender();
      }
      
/* 
 *  initm - initialize the merge process - open fb_input, check output... 
 */
 
   static initm(argc, argv, filen)
      int argc;
      char *argv[], *filen;
      
      {
         int i;
	 char err[FB_MAXLINE];
	 
	 filen[0] = NULL;
         linewidth = cdb_mergecols;
	 for (i = 1; i < argc; i++){
	    if (argv[i][0] == '-' && argv[i][1] == 'w'){
               i++;
               linewidth = atoi(argv[i]);
               }
	    else if (argv[i][0] == '-' && argv[i][1] == 'n'){
               i++;
               ncount = atoi(argv[i]);
               }
	    else if (argv[i][0] == '-' && argv[i][1] != NULL){
               switch(argv[i][1]){
                  case 'B': Bflag = 1;
                  case 'b':
                  case 'y':
                     break;
                  default:
                     i++;
                  }
	       }
	    else if (inpname[0] == NULL){
	       strcpy(inpname, argv[i]);
	       }
	    else if (filen[0] == NULL)
	       strcpy(filen, argv[i]);
	    }
         if (linewidth == 0)
            linewidth = 80;
	 if (cdb_batchmode && (strlen(filen) == 0 || strlen(inpname) == 0))
	    usage();
	 if (inpname[0] == NULL){
	    if (fb_getfilename(inpname, "Input File Name: ", "merge.in")==FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 if (filen[0] == NULL)
	    if (fb_getfilename(filen, "Output File Name: ", "merge.out")==FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
         if (strcmp(filen, inpname) == 0)
	    fb_xerror(FB_MESSAGE, SYSMSG[S_ILLEGAL], "Output/Input same file");
	 if (!cdb_batchmode && !cdb_yesmode)
	    if (access(filen, 0) != -1){
	       fb_move(4, 1);
	       fb_bell();
	       sprintf(err, 
		  "Permission to OVERWRITE %s (y=yes, <anyother>=no)? ",
		   filen);
	       if (fb_mustbe('y', err, 15, 10) == FB_ERROR)
		  fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	       }
         fb_readmrg(inpname);
	 mpcur = mphead;
         checkbegin();
         mpcur->mp_acur = mpcur->mp_ahead;
         mpcur->mp_atop = mpcur->mp_ahead;
      }
	 
/* 
 *  mparse - display list of tokens
 */
 
   static mparse(filen)
      char *filen;
      
      {
         int i, row;
         char *p, tbuf[FB_MAXLINE];
         fb_aline *a;
         fb_token *t;
	 
	 row = 1;
         fb_move(3,1);
         fb_printw("Input: %s   Output: %s", inpname, filen);
         fb_move(5, 1);
         fb_printw("Tokens: \n");
	 for (a = mpcur->mp_ahead; a != NULL; a = a->a_next){
            for (p = a->a_text, t = a->a_thead; *p; p++){
               if (*p == CHAR_DOLLAR && t != NULL){
                  if (t->t_field != NULL)
                     sprintf(tbuf, "$%s", t->t_field->id);
                  else
                     sprintf(tbuf, "$");
                  sprintf(tbuf, "%-10s[%2d]", tbuf, t->t_width);
                  /* display it */
                  if (row > SCREENLIMIT){
                     if (row <= SCREENLIMIT+1)
                        FB_PAUSE();
                     fb_move(6,1);
                     fb_clrtobot();
                     fb_infoline();
                     fb_move(6, 1);
                     row = 1;
                     }
	          fb_printw("   ... %s\n", tbuf);
	          row++;

                  /* move on to next token */
                  t = t->t_next;
                  }
               }
	    }
      }

/* 
 *  merge - merge an fb_input file across an indexed fb_database 
 */
 
   static merge(buf, filen)
      char *buf, filen[];
      
      {
         int mkpart();
         fb_aline *a;

         counter = 0L;
	 if (!cdb_batchmode){
            fb_scrstat("Merging");
	    FB_XWAIT();
	    fb_gcounter(counter);
	    }
	 if (equal(filen, "-"))
	    outfd = 1;
	 else{
	    close(creat(filen, 0666));
	    outfd = fb_mustopen(filen, WRITE);
            }
         /* fb_trim any white space out */
	 for (a = mpcur->mp_ahead; a != NULL; a = a->a_next)
            fb_trim(a->a_text);
	 for (a = a_begin; a != NULL; a = a->a_next)
            fb_trim(a->a_text);
	 for (a = a_end; a != NULL; a = a->a_next)
            fb_trim(a->a_text);
         /* initialize the output file */
         fb_w_init(1, outfd, 0);
	 /* print any BEGINSYMBOL stuff  */
	 for (a = a_begin; a != NULL; a = a->a_next){
            fb_nextwrite(0, a->a_text);
            fb_nextwrite(0, "\n");
            }
	 if (hp->dindex == NULL || hp->dindex[0] == NULL || hp->ihfd < 0){
            if (Bflag == 0)
               fb_foreach(hp, mkpart);
            else
               fb_blockeach(hp, mkpart);
            }
	 else
            fb_forxeach(hp, mkpart);
	 /* print any ENDSYMBOL stuff  */
	 for (a = a_end; a != NULL; a = a->a_next){
            fb_nextwrite(0, a->a_text);
            fb_nextwrite(0, "\n");
            }
         fb_wflush(1);
	 if (outfd != 1)
	    close(outfd);
         return(FB_AOK);
      }

/* 
 *  mkpart - make part of output for this buf (record) 
 */
 
   static mkpart(hp)
      fb_database *hp;
      
      {
         char *fld, *tmp, *p, *fb_formdollar(), *fb_rmlead();
         fb_aline *a;
         fb_token *t;
         fb_field *f;
         int width;
	 
         ++counter;
         /* test whether to force drop out of foreach/forxeach loop */
         if (ncount > 0 && counter > ncount)
            return(FB_ERROR);
         if (!cdb_batchmode)
            fb_gcounter(counter);
	 fld = cdb_afld; tmp = cdb_bfld;	/* set locals to globals */
         /* set current aline to place beyond BEGINSYMBOL */
	 for (a = mpcur->mp_ahead; a != NULL; a = a->a_next){
            for (p = a->a_text, t = a->a_thead; *p; p++){
               if (*p == CHAR_DOLLAR && t != NULL){
                  if (t->t_field != NULL){
                     f = t->t_field;
                     fb_fetch(f, fld, hp);
                     if ((width = t->t_width) == 0)
                        width = f->size;
                     fb_formfield(tmp, fld, f->type, width);
                     if (t->t_width > 0 && strlen(tmp) != t->t_width)
                        fb_pad(fld, tmp, t->t_width);
                     else
                        strcpy(fld, tmp);
                     }
                  else
                     strcpy(fld, "$");
                  fb_nextwrite(0, fld);
                  t = t->t_next;
                  }
               else
                  fb_w_write(0, p);
               }
            fb_w_write(0, "\n");
	    }
         return(FB_AOK);
       }

/*
 * checkbegin - check for BEGINSYMBOL and ENDSYMBOL stuff
 *       remove these sections from the list, use alone as text.
 */

   static checkbegin()

      {
         fb_aline *a;
         char *p;

	 for (a = mpcur->mp_ahead; a != NULL; a = a->a_next){
            p = a->a_text;
            if (strncmp(p, BEGINSYMBOL, 7) == 0){
               if (a_begin == NULL)
                  a_begin = a->a_next;
               else{
                  /* looking at last BEGINSYMBOL */
                  mpcur->mp_ahead = a->a_next;          /* patch new head */
                  a->a_next->a_prev = NULL;
                  a->a_prev->a_next = NULL;           /* patch begin list */
                  }
               }
            else if (strncmp(p, ENDSYMBOL, 5) == 0){
               if (a_end == NULL){
                  /* looking at first ENDSYMBOL */
                  a_end = a->a_next;
                  mpcur->mp_atail = a->a_prev;
                  a->a_prev->a_next = NULL;           /* patch tail of list */
                  a->a_prev = NULL;
                  }
               else{
                  /* looking at last ENDSYMBOL */
                  a->a_prev->a_next = NULL;           /* patch end list */
                  }
               }
            }
      }

/* 
 *  usage message 
 */
 
   usage()
      {
         fb_xerror(FB_MESSAGE, 
	    "usage: dbmerge [-d dbase] [-i index] [-b] [-y] input {output|-}",
	    NIL);
      }
